Lesson 14 (#42)
ReDI-School/nrw-html-and-css-2021-spring: lesson14/index.md

---
title: Lesson 14 - CSS Layout 2 - Flexbox
nav_order: 14
---

## Class Agenda

1. Introduction to Flexbox
1. The Flex model

## 1. Introduction to Flexbox

Flexbox is a comparatively newer way to layout elements on the page without the limits and frustration of Floats like:
  - You have to clear Floats (when the containing element collapses).

>   - Sometimes have to change the order of your HTML to achieve a Floats design

Reviewer: ```suggestion
  - Sometimes have to change the order of your HTML to achieve a Floats design.
```

Flexbox layout makes it easier to design flexible responsive layout structure (hint: Flexbox = flexible boxes).

> With flexbox, you can align elements horizontally and vertically. Unlike floats, you can re-order flex items without needing to change the HTML

Reviewer: ```suggestion
With flexbox, you can align elements horizontally and vertically. Unlike floats, you can re-order flex items without needing to change the HTML.
```

## 2. The Flex Model

- `flex` is the value for the display property that activates the Flexbox layout on an element.
- What is a flex container?
- What are flex items?

> - Explain some properties that can be used on a flex container

Reviewer: I find it a little bit non-fluent that we have a set of questions and somewhere in between the imperative sentence `Explain ....` 

```suggestion
- What are the most common properties that can be used on a flex container?
```

Author: yes, good point. Thanks

  - `flex-direction` (default `row`)
  - `flex-wrap` (default `nowrap`)
  - `justify-content` (default: `flex-start`)
  - `align-items` (default: `stretch`)
  - `align-content` (default: `stretch`)
  - `flex-flow`

> - Explain some properties that can be used on a flex item

Reviewer: ```suggestion
- What are the most common properties that can be used on a flex item?
```

>   - `align-self`, similar to `align-items` on the container, but applies to a single flex item.

Reviewer: ```suggestion
  - `align-self` is similar to `align-items` on the container, but applies to a single flex item.
```

  - `flex`: shorthand property for the `flex-grow`, `flex-shrink`, and `flex-basis`

>   - `flex-grow`: specifies how an item grows when they're allowed to take extra space available in the container (default: 0)

Reviewer: ```suggestion
  - `flex-grow`: specifies how an item grows when it is allowed to take extra space available in the container (default: 0)
```

  - `flex-shrink` specifies how an item shrinks when there is not enough space available in the container (default: 1)
  - flex-basis
  - order

# Exercise Description
1. Use Flexbox instead of float to style the navigation bar on the portfolio in [this CodeSandbox](https://codesandbox.io/s/flex1-vkyi2)

> 2. Use Flexbox to style the HTML form provided in [this CodeSandbox](https://codesandbox.io/s/flexbox2-ymdfj?file=/index.html) (see attached image for expected result)

Reviewer: Wooow good job here 💯  🚀 . Maybe we can use similar exercises for our exercise session (after making sure everyone catches up with the latest checkpoint)

Author: Yes. I had that idea too when doing this 😊

3. Style the Flexbox Pricing Grid in [this CodeSandbox](https://codesandbox.io/s/flexbox3-w8vqk) (see attached image for expected look)

# Homework

> - Checkpoint 14

Reviewer: ```suggestion
- Bring your website to Checkpoint 14.
```

> # Resources

Reviewer: ```suggestion
# Glossary

- Flexbox - Flexible box

# Resources
```
